src/engineio/transport_emitter.rs

```diff
@@ -0,0 +1,166 @@
+    fn set_on_close<F>(&mut self, function: F) -> Result<()>
+    where
+        F: Fn(()) + 'static + Sync + Send;
+}
+
+impl TransportEmitter {
+    /// Creates an instance of `Transport`.
+    pub fn new(tls_config: Option<TlsConnector>, opening_headers: Option<HeaderMap>) -> Self {
+        TransportEmitter {
+            transport: Arc::new(Mutex::new(Transports::new(tls_config, opening_headers))),
+            on_error: Arc::new(RwLock::new(None)),
+            on_open: Arc::new(RwLock::new(None)),
+            on_close: Arc::new(RwLock::new(None)),
+            on_data: Arc::new(RwLock::new(None)),
+            on_packet: Arc::new(RwLock::new(None)),
+        }
+    }
+
+    pub(super) fn upgrade_websocket(&mut self, address: String) -> Result<()> {
+        self.transport.lock()?.upgrade_websocket(address)
+    }
+
+    pub(super) fn upgrade_websocket_secure(&mut self, address: String) -> Result<()> {
+        self.transport.lock()?.upgrade_websocket_secure(address)
+    }
+}
+
+impl EventEmitter for TransportEmitter {
+    /// Registers an `on_open` callback.
+    fn set_on_open<F>(&mut self, function: F) -> Result<()>
+    where
+        F: Fn(()) + 'static + Sync + Send,
+    {
+        let mut on_open = self.on_open.write()?;
+        *on_open = Some(Box::new(function));
+        drop(on_open);
+        Ok(())
+    }
+
+    /// Registers an `on_error` callback.
+    fn set_on_error<F>(&mut self, function: F) -> Result<()>
+    where
+        F: Fn(String) + 'static + Sync + Send,
+    {
+        let mut on_error = self.on_error.write()?;
+        *on_error = Some(Box::new(function));
+        drop(on_error);
+        Ok(())
+    }
+
+    /// Registers an `on_packet` callback.
+    fn set_on_packet<F>(&mut self, function: F) -> Result<()>
+    where
+        F: Fn(Packet) + 'static + Sync + Send,
+    {
+        let mut on_packet = self.on_packet.write()?;
+        *on_packet = Some(Box::new(function));
+        drop(on_packet);
+        Ok(())
+    }
+
+    /// Registers an `on_data` callback.
+    fn set_on_data<F>(&mut self, function: F) -> Result<()>
+    where
+        F: Fn(Bytes) + 'static + Sync + Send,
+    {
+        let mut on_data = self.on_data.write()?;
+        *on_data = Some(Box::new(function));
+        drop(on_data);
+        Ok(())
+    }
+
+    /// Registers an `on_close` callback.
+    fn set_on_close<F>(&mut self, function: F) -> Result<()>
+    where
+        F: Fn(()) + 'static + Sync + Send,
+    {
+        let mut on_close = self.on_close.write()?;
+        *on_close = Some(Box::new(function));
+        drop(on_close);
+        Ok(())
+    }
+}
+
+impl Transport for TransportEmitter {
+    fn emit(&self, address: String, data: Bytes, is_binary_att: bool) -> Result<()> {
+        self.transport.lock()?.emit(address, data, is_binary_att)
+    }
+
+    fn poll(&mut self, address: String) -> Result<Bytes> {
+        self.transport.lock()?.poll(address)
+    }
+
+    fn name(&self) -> Result<String> {
```

Comment:
This could probably return a `&'static str`.

Reply:
done https://github.com/1c3t3a/rust-socketio/pull/52/commits/c7ff0a0511af614b6243cf73ffc80b4ed93a9e52